### PR TITLE
REFACTOR-#3780: Remove code duplication for `PandasOnDaskDataframe`

### LINKED ISSUE #3780: REFACTOR: remove redefinition of `_row_lengths` and `_column_widths` functions in `PandasOnDaskDataframe`.
Basic implementation should work for `Dask` engine. It already works for `Ray` engine.

Links to basic implementations: https://github.com/modin-project/modin/blob/d590de080dddb7ed715a58d57c602d96322c4dcf/modin/core/dataframe/pandas/dataframe/dataframe.py#L105
https://github.com/modin-project/modin/blob/d590de080dddb7ed715a58d57c602d96322c4dcf/modin/core/dataframe/pandas/dataframe/dataframe.py#L124

Links to `Dask` redefinitions:
https://github.com/modin-project/modin/blob/d590de080dddb7ed715a58d57c602d96322c4dcf/modin/core/execution/dask/implementations/pandas_on_dask/dataframe/dataframe.py#L47
https://github.com/modin-project/modin/blob/d590de080dddb7ed715a58d57c602d96322c4dcf/modin/core/execution/dask/implementations/pandas_on_dask/dataframe/dataframe.py#L64

### PR DESCRIPTION
**Pull request by anmyachev**
Signed-off-by: Anatoly Myachev <anatoly.myachev@intel.com>

## What do these changes do?

Now there is no need to override methods `_row_length` and `_column_width` for Dask separately, since the default implementation allows us to gather remote objects at once https://github.com/modin-project/modin/issues/3780#issuecomment-1234177909

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #3780 
- [x] tests ~added~ and passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date

### CODE REVIEW

**Review by vnlitvinov — COMMENTED** [review]
LGTM, but did you run any benchmarks? While general code might work, it might do so slower.

**Comment by anmyachev** [comment]
> LGTM, but did you run any benchmarks? While general code might work, it might do so slower.

No. 
The mechanism for obtaining metadata differs only in one thing: we receive all the results at one time (via `gather`), or we should make several calls (via `result` for every partition), but everywhere we use lambdas as recommended by [Dask](http://distributed.dask.org/en/stable/efficiency.html). So the performance shouldn't be different.

**Comment by vnlitvinov** [comment]
I would rather saw this proved, rather than believe in the docs... I saw too much docs being liars to believe anything but firm fact measured by someone I can trust (like you 🙃)

**Review by devin-petersohn — COMMENTED** [review]
Removing this will probably make things slower. Instead of gathering all at once, this will get each one individually.

**Comment by dchigarev** [comment]
> Removing this will probably make things slower.

But not for all of the cases. Partitions have their own cache for [`width`](https://github.com/modin-project/modin/blob/cf426c4839818667dc4cc5349d4f3c834a602877/modin/core/execution/dask/implementations/pandas_on_dask/partitioning/partition.py#L321-L325) and [`length`](https://github.com/modin-project/modin/blob/cf426c4839818667dc4cc5349d4f3c834a602877/modin/core/execution/dask/implementations/pandas_on_dask/partitioning/partition.py#L306-L310). If the cache is filled, this will speed up things instead.

**Comment by anmyachev** [comment]
> > Removing this will probably make things slower.
> 
> But not for all of the cases. Partitions have their own cache for [`width`](https://github.com/modin-project/modin/blob/cf426c4839818667dc4cc5349d4f3c834a602877/modin/core/execution/dask/implementations/pandas_on_dask/partitioning/partition.py#L321-L325) and [`length`](https://github.com/modin-project/modin/blob/cf426c4839818667dc4cc5349d4f3c834a602877/modin/core/execution/dask/implementations/pandas_on_dask/partitioning/partition.py#L306-L310). If the cache is filled, this code removal will speed up things instead.

Good point. So the best solution here would be to take all the lengths at a time, through the appropriate engine interface. After which it will be necessary to fill the cache of separate partitions in a loop.

**Review by devin-petersohn — COMMENTED** [review]
Can we just add a check for the cache into these functions engine-specific? The lengths/widths cache should usually be there, but in the case it's not we'll see very expensive operation times.

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 1]
```diff
@@ -231,7 +236,16 @@ def _column_widths(self):
...
+                partitions = self._partitions[0]
+                if all(map(lambda t: t._width_cache is not None, partitions)):
+                    self._column_widths_cache = [obj.width() for obj in partitions]
+                else:
+                    widths = [
+                        obj.apply(lambda df: len(df.columns)) for obj in partitions
+                    ]
+                    self._column_widths_cache = (
+                        self._partition_mgr_cls.get_objects_from_partitions(widths)
+                    )
```
Comment (dchigarev) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:248]: shouldn't we handle this at the PartitionClass level?

Comment (dchigarev) [reply modin/core/dataframe/pandas/dataframe/dataframe.py:248]: the code also makes to recompute cache for ALL of the partitions if only ONE of them has missing cache

**Comment by anmyachev** [comment]
@YarShev @vnlitvinov @mvashishtha please review

**Review by mvashishtha — APPROVED** [review]
LGTM @anmyachev ! Thank you.

**Review by vnlitvinov — APPROVED** [review]
LGTM, nice find!

**Comment by anmyachev** [comment]
@mvashishtha let's merge it

**Pull request merged**